### PR TITLE
Fix typos and grammar in the OTLP spec

### LINKED ISSUE #3120: Fix typos and grammar in the OTLP specification
**What are you trying to achieve?**

There are several typos and grammatical errors in the OTLP spec. There are several places where the text can be re-written to make the ideas more clear and concise.

What did you expect to see?

I expect to see minimal (even no) typos/grammatical errors.

### PR DESCRIPTION
**Pull request by moonbox3**
Fixes #3120 

## Changes

This PR fixes typos and grammatical errors in the OTLP spec. Additionally, there were several places where text contained passive voice. These were updated to active voice to make them easier to follow.

### CODE REVIEW

**Comment by moonbox3** [comment]
Does this change require a changelog?

File: specification/protocol/otlp.md [hunk 1]
```diff
@@ -648,50 +652,51 @@ client-side queue).
...
+etc.) whether recently provided data was delivered and whether an acknowledgment 
+was received or not. Therefore, the client will typically choose to re-send such 
+data to guarantee delivery, which may result in duplicate data on the server side. 
+This is a deliberate choice and is considered to be the right tradeoff for telemetry 
+data.
 
 ## Future Versions and Interoperability
 
 OTLP will evolve and change over time. Future versions of OTLP must be designed
-and implemented in a way that ensures that clients and servers that implement
```
Comment (tigrannajaryan) [inline specification/protocol/otlp.md:660]: What's wrong with "in a way that"?

**Review by tigrannajaryan — CHANGES_REQUESTED** [review]
Thank you for the fixes @moonbox3

Some of the changes DO change the meaning, so I am temporarily blocking to fix those.

File: specification/protocol/otlp.md [hunk 2]
```diff
@@ -97,27 +97,27 @@ to receive a response to each request:
...
 acknowledgements described in this protocol happen between a single
 client/server pair and do not span intermediary nodes in multi-hop delivery
 paths._
 
 #### OTLP/gRPC Concurrent Requests
 
-After sending the request the client MAY wait until the response is received
-from the server. In that case there will be at most only one request in flight
-that is not yet acknowledged by the server.
+After sending the request, the client MAY wait until the server receives a response.
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply specification/protocol/otlp.md:112]: Will fix.

File: specification/protocol/otlp.md [hunk 3]
```diff
@@ -97,27 +97,27 @@ to receive a response to each request:
...
+In that case, there will be, at most only one request in flight that the server does 
+not yet acknowledge.
 
 ![Unary](img/otlp-sequential.png)
 
 Sequential operation is recommended when simplicity of implementation is
-desirable and when the client and the server are connected via very low-latency
-network, such as for example when the client is an instrumented application and
+desirable, and when the client and the server are connected via very low-latency
+network, such as when the client is an instrumented application and
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply specification/protocol/otlp.md:120]: "Such as" and "for example" are conveying the same thing.

File: specification/protocol/otlp.md [hunk 4]
```diff
@@ -97,27 +97,27 @@ to receive a response to each request:
...
 paths._
 
 #### OTLP/gRPC Concurrent Requests
 
-After sending the request the client MAY wait until the response is received
-from the server. In that case there will be at most only one request in flight
-that is not yet acknowledged by the server.
+After sending the request, the client MAY wait until the server receives a response. 
+In that case, there will be, at most only one request in flight that the server does 
+not yet acknowledge.
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply specification/protocol/otlp.md:114]: I'm trying to get away from passive voice, which the original was written in.

[the remark this answers is not in the document]
Comment (moonbox3) [reply specification/protocol/otlp.md:114]: Reverted to the original.

File: specification/protocol/otlp.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (moonbox3) [reply specification/protocol/otlp.md:120]: You don't agree that "such as for example when the client is an instrumented application..." can be reduced to read better as: "such as when the client is an instrumented application..." ?

File: specification/protocol/otlp.md (hunk 1 above)
Comment (moonbox3) [reply specification/protocol/otlp.md:660]: "In a way that ensures" felt too wordy so I wanted to rewrite it for clarity. I really want it to be "... OTLP must be designed and implemented to ensure that clients..." Do you agree?

Comment (moonbox3) [reply specification/protocol/otlp.md:660]: I will just revert it...

**Comment by moonbox3** [comment]
> Thank you for the fixes @moonbox3
> 
> Some of the changes DO change the meaning, so I am temporarily blocking to fix those.

Thank you for your review, @tigrannajaryan. I appreciate it. I have addressed your comments/suggestions.

**Comment by moonbox3** [comment]
> Btw we don't use comma after `e.g.` nor `i.e.` in general - according to [Wiktionary](https://en.wiktionary.org/wiki/e.g.):
> 
> > "E.g." and its examples are typically set off from the rest of the sentence by punctuation. In US English, "e.g." is usually followed by a comma.[[2]](https://en.wiktionary.org/wiki/e.g.#cite_note-2) It is not followed by a comma in other English-speaking countries.
> 
> So we should probably stick to the comma-less form for now, given we already use that and, _technically_, it's not incorrect.

@carlosalberto Thank you for your comment. Is there a style guide I should be following?

File: specification/protocol/otlp.md [hunk 5]
```diff
@@ -235,17 +235,17 @@ containing 0 value of RetryDelay. Here is a sample Go code to illustrate:
...
-To indicate not-retryable errors the server is recommended to use code
+To indicate not-retryable errors, the server is recommended to use code
 [InvalidArgument](https://godoc.org/google.golang.org/grpc/codes) and MAY supply
 additional
 [details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails)
 using
 [BadRequest](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L119).
-Other gRPC status code may be used if it is more appropriate. Here is a sample
-Go code to illustrate:
+If more appropriate, another gRPC status code MAY be used. Here is a
```
Comment (MrAlias) [inline specification/protocol/otlp.md:244]: This changes "may" into a key word and changes semantic meaning. Is that intentional?

File: specification/protocol/otlp.md [hunk 6]
```diff
@@ -189,31 +189,31 @@ the number of spans/data points/log records it rejected.
...
 ##### Failures
 
-When an error is returned by the server it falls into 2 broad categories:
+When the server returns an error, it falls into 2 broad categories:
 retryable and not-retryable:
 
-- Retryable errors indicate that processing of telemetry data failed and the
-  client SHOULD record the error and may retry exporting the same data. This can
+- Retryable errors indicate that telemetry data processing failed, and the
+  client SHOULD record the error and may retry exporting the same data. For example, this can
```
Comment (MrAlias) [inline specification/protocol/otlp.md:210]: This should follow the same wrapping length as the rest of the paragraph.

File: specification/protocol/otlp.md [hunk 7]
```diff
@@ -189,31 +189,31 @@ the number of spans/data points/log records it rejected.
...
   happen when the server is temporarily unable to process the data.
 
-- Not-retryable errors indicate that processing of telemetry data failed and the
-  client MUST NOT retry sending the same telemetry data. The telemetry data MUST
-  be dropped. This can happen, for example, when the request contains bad data
-  and cannot be deserialized or otherwise processed by the server. The client
+- Not-retryable errors indicate that telemetry data processing failed, and the
+  client MUST NOT retry sending the same telemetry data. The client MUST drop the telemetry data.
```
Comment (MrAlias) [inline specification/protocol/otlp.md:214]: This should follow the same wrapping length as the rest of the paragraph.

File: specification/protocol/otlp.md [hunk 8]
```diff
@@ -500,7 +504,7 @@ in case of a successful response.
 ##### Partial Success
 
 If the request is only partially accepted
-(i.e. when the server accepts only parts of the data and rejects the rest), the
+(i.e., when the server accepts only parts of the data and rejects the rest), the
```
Comment (MrAlias) [inline specification/protocol/otlp.md:507]:
```suggestion
(i.e. when the server accepts only parts of the data and rejects the rest), the
```

File: specification/protocol/otlp.md [hunk 9]
```diff
@@ -575,29 +580,29 @@ response.
...
+The client SHOULD honour the waiting interval specified in the "Retry-After" header
+if it is present. If the client receives an `HTTP 429` or an `HTTP 503` response and
+the "Retry-After" header is not present in the response, then the client SHOULD
 implement an exponential backoff strategy between retries.
 
 ##### All Other Responses
 
 All other HTTP responses that are not explicitly listed in this document should
-be treated according to HTTP specification.
+be treated according to HTTP specifications.
```
Comment (MrAlias) [inline specification/protocol/otlp.md:596]: Are there multiple HTTP specifications?

File: specification/protocol/otlp.md [hunk 10]
```diff
@@ -625,15 +630,14 @@ The default network port for OTLP/HTTP is 4318.
...
-forward.
+An additional complication must be accounted for when one client must send 
+telemetry data to more than one destination server. When one of the servers 
+acknowledges the data and the other server does not (yet), the client needs 
+to decide how to move forward.
 
-In such situation the client SHOULD implement queuing, acknowledgement
-handling and retrying logic per destination. This ensures that servers do not
-block each other. The queues SHOULD reference shared, immutable data to be sent,
+In such a situation, the client SHOULD implement queuing, acknowledgment-handling
```
Comment (MrAlias) [inline specification/protocol/otlp.md:638]: Adding the leading comma here makes the second sentence fragment appear as a contextual element. Which does not seem to be the original intent of the specification.

File: specification/protocol/otlp.md [hunk 11]
```diff
@@ -97,11 +97,11 @@ to receive a response to each request:
...
 ![Request-Response](img/otlp-request-response.png)
 
-_Note: this protocol is concerned with reliability of delivery between one pair
-of client/server nodes and aims to ensure that no data is lost in-transit
+_Note: this protocol is concerned with the reliability of delivery between one
+pair of client/server nodes and aims to ensure that no data is lost in transit
 between the client and the server. Many telemetry collection systems have
 intermediary nodes that the data must travel across until reaching the final
-destination (e.g. application -> agent -> collector -> backend). End-to-end
+destination (e.g., application -> agent -> collector -> backend). End-to-end
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply specification/protocol/otlp.md:104]: It seems it would be a good idea to add the "no comma needed after e.g." to some style guide in this project.

File: specification/protocol/otlp.md (hunk 5 above)
Comment (moonbox3) [reply specification/protocol/otlp.md:244]: Didn't mean for may to be capitalized. Fixing.

File: specification/protocol/otlp.md (hunk 9 above)
Comment (moonbox3) [reply specification/protocol/otlp.md:596]: Sure, there exists more than one "specification." As an example, see: https://www.w3.org/Protocols/Specs.html

File: specification/protocol/otlp.md (hunk 10 above)
Comment (moonbox3) [reply specification/protocol/otlp.md:638]: I don't agree that the comma makes the second sentence fragment appear as a contextual element. "In such situation" is an example of an introductory phrase or clause, which sets the stage for a sentence. A writer may use discretion to determine if the comma is needed.

Comment (MrAlias) [reply specification/protocol/otlp.md:638]: I think both can be true.

Is there a way to restructure this sentence to remove confusion and clarify the meaning? I'm still not 100% sure if this is a list of things that needs to be addressed "per destination" or if only the retry logic is "per destination".

Comment (moonbox3) [reply specification/protocol/otlp.md]: I think that's a valid callout. This text is under the heading "Multi-Destination Exporting" so I understand it as:

"The client SHOULD implement (per destination): queuing, acknowledgement-handling, and retry logic."

**Comment by moonbox3** [comment]
@tigrannajaryan or @MrAlias can you please help review the final changes? Thank you.

**Comment by moonbox3** [comment]
I would prefer for this PR to not be closed due to lack of activity. Can I get some help on it, please? Thank you.

**Review by MrAlias — APPROVED** [review]

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
Thanks @moonbox3 - looks good to go IMHO. 

cc @tigrannajaryan

File: specification/protocol/otlp.md [hunk 12]
```diff
@@ -328,30 +334,30 @@ Here is a sample Go code to illustrate:
...
-// number of retires have been reached or a maximum retry delay cap has been
+// fail, clients should use an exponential backoff scheme to increase gradually
+// the delay between retries based on `retry_delay` until either a maximum
+// number of retries has been reached, or a maximum retry delay cap has been
 // reached.
 ```
 
-The value of `retry_delay` is determined by the server and is implementation
+The server determines the value of the `retry_delay` and is implementation
 dependant. The server SHOULD choose a `retry_delay` value that is big enough to
```
Comment (tigrannajaryan) [inline specification/protocol/otlp.md:358]: >The server determines the value of the `retry_delay` and is implementation
>dependant.

Unclear what the `is` in the second part of the sentence refers to (the server? the value?).

File: specification/protocol/otlp.md [hunk 13]
```diff
@@ -625,15 +631,14 @@ The default network port for OTLP/HTTP is 4318.
...
+An additional complication must be accounted for when one client must send
+telemetry data to more than one destination server. When one of the servers
+acknowledges the data and the other server does not (yet), the client needs
+to decide how to move forward.
 
-In such situation the client SHOULD implement queuing, acknowledgement
-handling and retrying logic per destination. This ensures that servers do not
-block each other. The queues SHOULD reference shared, immutable data to be sent,
+In such a situation, the client SHOULD implement (per destination): queuing,  
+acknowledgment-handling, and retry logic. This ensures that servers do not
```
Comment (tigrannajaryan) [inline specification/protocol/otlp.md:640]: >In such a situation, the client SHOULD implement (per destination): queuing,  
>acknowledgment-handling, and retry logic.

I think this changed the meaning of what we want to say. The emphasis here is that all these capabilities must be implemented per destination, not that these capabilities must be implemented. Putting "per-destination" in parenthesis no longer conveys that important message.

**Review by tigrannajaryan — COMMENTED** [review]
Just 2 small comments left, the rest looks good to me.

File: specification/protocol/otlp.md (hunk 12 above)
Comment (moonbox3) [reply specification/protocol/otlp.md:358]: I will revert to the original if it is confusing. I understand the "is" is referring to the value of the retry delay.

File: specification/protocol/otlp.md (hunk 13 above)
Comment (moonbox3) [reply specification/protocol/otlp.md:640]: I reverted to the original.

**Comment by moonbox3** [comment]
Thank you all for the help. @tigrannajaryan I addressed your final two comments. Thank you.

**Review by tigrannajaryan — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
Thank you @moonbox3 for making the spec clearer.
